library/src/main/java/io/ona/kujaku/services/TrackingService.java

```diff
@@ -0,0 +1,939 @@
+            processLocation(location);
+        }
+
+    };
+
+    /**
+     * Service thread with looper to handle GPS notification Thread to process all notifications
+     */
+    private volatile Thread serviceThread = new Thread("TrackingService") {
+        public void run() {
+            Log.d(TAG, "Tracking thread started.");
+            // preparing a looper on current thread
+            // the current thread is being detected implicitly
+            Looper.prepare();
+
+            Log.d(TAG, "Register GPS status listener.");
+
+            // No need to do it in thread as the listener only logs
+            // which is fast
+            // locationManager.addGpsStatusListener(gpsListener);
+
+            // now, the handler will automatically bind to the
+            // Looper that is attached to the current thread
+            // You don't need to specify the Looper explicitly
+            gpsHandler = new Handler();
+
+            // Register the Location listener
+            registerLocationListener();
+
+            // After the following line the thread will start
+            // running the message loop and will not normally
+            // exit the loop unless a problem happens or you
+            // quit() the looper (see below)
+            Looper.loop();
+
+            Log.d(TAG, "Exiting looper.");
+
+            if (pendingRecordingLocation != null) {
+
+                Log.d(TAG, "Record last pending location.");
+                recordPendingLocation();
+            }
+
+            // Mark to notify thread is exiting
+            serviceThreadRunningLatch.countDown();
+        }
+    };
+
+    /**
+     * Call by service Thread to stop itself
+     */
+    private Runnable stopServiceThread = new Runnable() {
+        @Override
+        public void run() {
+            Looper looper = Looper.myLooper();
+            if (looper != null) {
+                looper.quit();
+            } else {
+                Log.e(TAG, "Cannot stop service thread.");
+            }
+        }
+    };
+
+    /**
+     * Get Wake Lock
+     *
+     * @return
+     */
+    private PowerManager.WakeLock getWakeLock() {
+        if (this.wakeLock == null) {
+            // Create the wakeLock to prevent the device from sleeping
+            this.wakeLock = this.powerManager.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK,
+                    PARTIAL_WAKE_LOCK_TAG);
+        }
+
+        return this.wakeLock;
+    }
+
+    /**
+     * To ensure the service is not killed too easily
+     */
+    protected void startServiceForeground() {
+        String channel = "";
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            channel = createChannel();
+        }
+
+        NotificationCompat.Builder mBuilder = new NotificationCompat.Builder(this, channel)
+                .setSmallIcon(android.R.drawable.ic_menu_mylocation)
+                .setContentTitle(getString(R.string.tracking_service_notification_title))
```

Comment:
Since this is going to be used in client applications with different names, can we make this either configurable, use a general name or include the declared app name instead? Does it make sense?

Reply:
Yes you're right, i will it change to use the app name

Reply:
Sorry, I meant `%1$s Tracking Service` and have the `%1$s` filled by the app name. You can use `String.format`. Use the app name from the resources us you have done above.

Reply:
understood